crossbar/worker/proxy.py

```diff
@@ -89,352 +363,140 @@ def onConnect(self):
         pass
 
     def onClose(self, wasClean):
-        if self._frontend.transport:
-            self._frontend.transport.loseConnection()
+        if self._frontend:
+            self._frontend.transport.send(message.Goodbye())
         self._frontend = None
         super(BackendProxySession, self).onClose(wasClean)
 
     def onMessage(self, msg):
         # 'msg' is a real WAMP message that our backend WAMP protocol
         # has deserialized -- so now we re-serialize it for whatever
         # the frontend is speaking
-        self._frontend.forward_message(msg)
-
-        # ... should we do this explicitly, or just count on the
-        # "disconnect" propogating properly
-        if False and isinstance(msg, Goodbye) and self._frontend:
-            self._frontend.transport.loseConnection()
-            self._frontend = None
+        # print("BackendProxySession.onMessage: {}".format(msg))
+        if isinstance(msg, (message.Welcome, message.Challenge, message.Abort, message.Goodbye)):
```

Comment:
exactly! this is the crucial code that defines the behavior of the new WAMP proxy:

* it handles frontend and backend wamp opening handshake / authentication, but _not_
* session-level wamp message types, which it instead forwards directly (without any processing at the app level, but with re-serialization if required)